File: src/106.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    '''
    大师，我悟了，中序用来判断左右，后序从后向前获得节点
    '''
    '''
    仍然递归
    1. 根据后序序列的最后一个元素建立根结点；
    2. 在中序序列中找到该元素，确定根结点的左右子树的中序序列；
    3. 在后序序列中确定左右子树的后序序列；
    4. 由左子树的后序序列和中序序列建立左子树；
    5. 由右子树的后序序列和中序序列建立右子树
    '''
    '''
    time debeat: 14%
    space debeat: 11%
    '''
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(postorder) == 0:
            return None
        elif len(postorder) ==1:
            return TreeNode(postorder.pop())
        elif len(postorder) ==2:
            #当只有2个时，构造一棵树回去
            #根节点
            root = TreeNode(postorder.pop())
            #while len(postorder) > 0:
            #构造下级节点
            newNode = TreeNode(postorder.pop())
            #根据左右构造路径 
            if inorder.index(newNode.val) < inorder.index(root.val):
                root.left = newNode
            else:
                root.right = newNode
            return root
        else:
            #根节点 
            root = TreeNode(postorder[-1])
            #右子中序和后序
            rightInOrder = inorder[inorder.index(root.val) + 1 :]
            rightPostOrder = postorder[-1 - len(rightInOrder) : -1]
            root.right = self.buildTree(rightInOrder, rightPostOrder)
            #左子中序和后序
            leftInOrder = inorder[:inorder.index(root .val)]
            leftPostOrder = postorder[: -1 - len(rightInOrder)]
            root.left = self.buildTree(leftInOrder, leftPostOrder)
            return root
        

File: src/106.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode * constructTree(const vector<int> & _inorder, int _iLeft, int _iRight, const vector<int> & _posterorder, int _pLeft, int _pRight){
        //递归构造
        //递归边界
        if(_iLeft > _iRight){
            return nullptr;
        }
        //先构造根节点
        TreeNode * node = new TreeNode(_posterorder[_pRight]);
        //区分左右子树
        //题目保证数字不重复
        int nodeIndex = find(_inorder.begin(), _inorder.end(), _posterorder[_pRight]) - _inorder.begin();
        int leftSubTreeLength = nodeIndex - _iLeft;
        //cout<<_posterorder[_pRight]<<_iLeft<<_iRight<<_pLeft<<_pRight<<endl;
        //向下构造
        node->left = constructTree(_inorder, _iLeft, _iLeft + leftSubTreeLength - 1, _posterorder, _pLeft, _pLeft + leftSubTreeLength - 1);
        node->right = constructTree(_inorder, _iLeft + leftSubTreeLength + 1, _iRight, _posterorder, _pLeft + leftSubTreeLength, _pRight - 1);
        return node;
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        //同样的构建方法
        //干翻蚂蚁链
        //中序遍历，[左子树，根，右子树]
        //后序遍历[左，右，根]
        //通过参数指针，避免参数出现多次拷贝
        //草了，因为一个参数卡了半天，
        //二十五分钟，一次提交，时间击败--88.91%，空间击败-74.79%
        TreeNode * root = constructTree(inorder, 0, inorder.size() - 1, postorder, 0, postorder.size() - 1);
        return root;
    }
};
